### PR TITLE
adds Gaussian NLL Loss

### LINKED ISSUE #48520: Gaussian NLL loss
## 🚀 Feature
Gaussian negative log-likelihood loss, similar to issue #1774 (and solution pull #1779) 

## Motivation

The homoscedastic Gaussian loss is described in Equation 1 of this [paper](https://arxiv.org/pdf/1612.01474.pdf). The heteroscedastic version in Equation 2 [here](https://arxiv.org/pdf/2011.02838.pdf)  (ignoring the final anchoring loss term). These are both key to the uncertainty quantification techniques described. 

## Pitch

I'm happy to implement this, using the template of pull #1779. The implementation will allow for both homoscedastic and heteroscedastic losses.  

## Alternatives

An alternative would be to instantiate a Gaussian (https://pytorch.org/docs/stable/distributions.html#normal) and evaluate the log of this. However, this seems wasteful given a new Gaussian would be instantiated for every new function call for the best case (homoscedastic), and for every element of the output-target pairs in the worst case (heteroscedastic). 

## Additional context

Definitions: [homo/heteroscedasticity](https://en.wikipedia.org/wiki/Homoscedasticity) (wiki)

cc @albanD @mruberry

### PR DESCRIPTION
**Pull request by nailimixaM**
Fixes #48520

cc @albanD @sparkingdark

To do:
- I need help with writing the equation in the GaussianNLLLoss class docstring so that it appears correctly in the docs. Is there a website where I can draft it, then paste it in?
- More tests, once the code is accepted

Noteworthy:
- eps in GaussianNLLLoss class: this is added to var to avoid tiny values of var. However, it assumes var has strictly non-negative entries, which should be tested for assuming this method is acceptable
- the loss function accepts input/target tensors of shape (N, M) and will also work for shape (N, M, 1) but *not* (N, 1, M) due to a torch.sum(..., dim=1) and length=input.shape[1] in the loss calculation. Is there a neat to fix this issue, or should (N, 1, M) simply not be accepted?

I have likely missed other issues, so thanks for the help in sorting these and the above!

### CODE REVIEW

File: torch/nn/modules/loss.py [hunk 1]
```diff
@@ -297,6 +297,38 @@ def forward(self, log_input: Tensor, target: Tensor) -> Tensor:
...
 
 
+class GaussianNLLLoss(_Loss):
+    r"""Negative log likelihood loss with Gaussian distribution
+
+    The loss can be described as::
+
+    target ~ N(input, var)
+
+    loss(input, target, var) = %TODO - need help with math
```
Comment (albanD) [inline torch/nn/modules/loss.py:307]: You can do something similar to the kldiv below:
```
    .. math::
        l(x,y) = L = \{ l_1,\dots,l_N \}, \quad
        l_n = y_n \cdot \left( \log y_n - x_n \right)
```
The syntax inside the `.. math::` construct is latex formula syntax IIRC

**Review by albanD — COMMENTED** [review]
Thanks for sending a PR!

I added comments inline.

File: torch/nn/functional.py [hunk 2]
```diff
@@ -2316,6 +2316,70 @@ def poisson_nll_loss(input, target, log_input=True, full=False, size_average=Non
...
+    #Calculate loss (without constant)
+    loss = 0.5*(torch.sum(torch.log(var) + (input - target)**2/var, dim=1))
+
+    #Add constant to loss term if required
+    if full:
+        length = input.shape[1]
+        loss = loss + 0.5*length*math.log(2*math.pi)
+
+    #Apply reduction
+    if reduction == 'mean':
```
[the remark this answers is not in the document]
Comment (albanD) [reply torch/nn/functional.py:2375]: Sounds good! Thanks for checking.

**Comment by nailimixaM** [comment]
@albanD I'll open a new PR, I think I messed this one up by attempting to rebase, sorry!
